src/core/socket-bind.c

```diff
@@ -0,0 +1,180 @@
+                int map_fd, CGroupSocketBindItem *head, enum socket_bind_action action, uint32_t* index) {
+        CGroupSocketBindItem *item;
+
+        assert(map_fd >= 0);
+        assert(index);
+
+        LIST_FOREACH(socket_bind_items, item, head) {
+                const uint32_t key = (*index)++;
+                struct socket_bind_rule val = {
+                        .address_family = (uint32_t) item->address_family,
+                        .nr_ports = item->nr_ports,
+                        .port_min = item->port_min,
+                        .action = action,
+                };
+
+                if (bpf_map_update_elem(map_fd, &key, &val, BPF_ANY) != 0)
+                        return -errno;
+        }
+
+        return 0;
+}
+
+static int prepare_socket_bind_bpf(
+                Unit *u, CGroupSocketBindItem *allow, CGroupSocketBindItem *deny, struct socket_bind_bpf **ret_obj) {
+        _cleanup_(socket_bind_bpf_freep) struct socket_bind_bpf *obj = 0;
+        uint32_t rule_count = 0, index = 0;
+        CGroupSocketBindItem *item;
+        int map_fd, r;
+
+        assert(ret_obj);
+
+        LIST_FOREACH(socket_bind_items, item,  allow)
+                rule_count += 1;
+
+        LIST_FOREACH(socket_bind_items, item, deny)
+                rule_count += 1;
+
+        if (rule_count > socket_bind_max_rules)
+                return log_unit_error_errno(u, SYNTHETIC_ERRNO(EINVAL),
+                                "Maximum number of socket bind rules=%u is exceeded", socket_bind_max_rules);
+
+        obj = socket_bind_bpf__open();
+        if (!obj)
+                return log_unit_error_errno(u, SYNTHETIC_ERRNO(ENOMEM), "Failed to open BPF object");
+
+        if (bpf_map__resize(obj->maps.rules, MAX(rule_count, 1u)) != 0)
+                return log_unit_error_errno(u, errno,
+                                "Failed to resize BPF map '%s': %m", bpf_map__name(obj->maps.rules));
+
+        if (socket_bind_bpf__load(obj) != 0)
+                return log_unit_error_errno(u, errno, "Failed to load BPF object");
+
+        map_fd = bpf_map__fd(obj->maps.rules);
+        assert(map_fd >= 0);
+
+        r = update_rules_map(map_fd, allow, SOCKET_BIND_ALLOW, &index);
+        if (r < 0)
+                return log_unit_error_errno(
+                                u, r, "Failed to put socket bind allow rules into BPF map '%s'",
+                                bpf_map__name(obj->maps.rules));
+
+        r = update_rules_map(map_fd, deny, SOCKET_BIND_DENY, &index);
+        if (r < 0)
+                return log_unit_error_errno(
+                                u, r, "Failed to put socket bind deny rules into BPF map '%s'",
+                                bpf_map__name(obj->maps.rules));
+
+        *ret_obj = TAKE_PTR(obj);
+        return 0;
+}
+
+int socket_bind_supported(void) {
+        _cleanup_(socket_bind_bpf_freep) struct socket_bind_bpf *obj = NULL;
+
+        int r = cg_unified_controller(SYSTEMD_CGROUP_CONTROLLER);
```

Comment:
@keszybz 
IIUC caching is already done in https://github.com/systemd/systemd/blob/main/src/basic/cgroup-util.c#L2092 ?

Reply:
I meant the result of `socket_bind_supported()`.
It's called from bus_cgroup_set_property(), so it could be called repeatedly...

Reply:
Looks like calling `socket_bind_supported()` in dbus-cgroup is not required after all since unit manager already has that info stored in `cgroup_supported`. It is being set in [manager_setup_cgroup](https://github.com/systemd/systemd/blob/main/src/core/cgroup.c#L3070).
 
Replaced `socket_bind_supported()` with 
`(u->manager->cgroup_supported & CGROUP_MASK_BPF_SOCKET_BIND) == 0`